File: plugin/src/test/java/com/oldstats/tracking/DiaryTaskTrackerTest.java
package com.oldstats.tracking;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.oldstats.api.OldStatsApiClient;
import com.oldstats.api.StatEvent;
import java.util.List;
import java.util.stream.Collectors;
import net.runelite.api.Client;
import net.runelite.api.events.WidgetLoaded;
import net.runelite.api.gameval.InterfaceID;
import net.runelite.api.widgets.Widget;
import net.runelite.client.callback.ClientThread;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

/**
 * DiaryTaskTracker reads the diary journal's "journal scroll" interface
 * (group InterfaceID.JOURNALSCROLL) rather than a varbit, so these tests
 * build fake title/text-layer widgets the same shape RuneLite's own
 * decompiled DiaryRequirementsPlugin reads (confirmed via cfr).
 */
public class DiaryTaskTrackerTest {
    private OldStatsApiClient apiClient;
    private Client client;
    private ClientThread clientThread;
    private DiaryTaskTracker tracker;

    @Before
    public void setUp() {
        apiClient = mock(OldStatsApiClient.class);
        client = mock(Client.class);
        clientThread = mock(ClientThread.class);
        tracker = new DiaryTaskTracker(apiClient, client, clientThread);

        // Run the deferred client-thread callback synchronously so tests can assert immediately.
        doAnswer(invocation -> {
            invocation.<Runnable>getArgument(0).run();
            return null;
        }).when(clientThread).invokeLater(any(Runnable.class));
    }

    private Widget widget(String text) {
        Widget w = mock(Widget.class);
        when(w.getText()).thenReturn(text);
        return w;
    }

    private void setPage(String heading, String... lines) {
        Widget titleWidget = widget("Achievement Diary");
        when(client.getWidget(InterfaceID.Journalscroll.TITLE)).thenReturn(titleWidget);

        Widget[] lineWidgets = new Widget[lines.length + 1];
        lineWidgets[0] = widget(heading);
        for (int i = 0; i < lines.length; i++) {
            lineWidgets[i + 1] = widget(lines[i]);
        }
        Widget textLayer = mock(Widget.class);
        when(textLayer.getStaticChildren()).thenReturn(lineWidgets);
        when(client.getWidget(InterfaceID.Journalscroll.TEXTLAYER)).thenReturn(textLayer);
    }

    private WidgetLoaded loadEvent() {
        return loadEvent(InterfaceID.JOURNALSCROLL);
    }

    private WidgetLoaded loadEvent(int groupId) {
        WidgetLoaded event = new WidgetLoaded();
        event.setGroupId(groupId);
        return event;
    }

    @Test
    public void emitsOneEventPerTaskLineWithTheCorrectTierAndCompletionState() {
        setPage(
            "Ardougne Area Tasks",
            "Easy",
            "Enter the Wilderness",
            "<str>Talk to the Ardougne baker</str>",
            "Medium",
            "Steal from the Ardougne market stalls"
        );

        tracker.onWidgetLoaded(loadEvent());

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient, times(3)).enqueue(captor.capture());
        List<StatEvent.DiaryTaskProgress> events = captor.getAllValues().stream()
            .map(e -> (StatEvent.DiaryTaskProgress) e)
            .collect(Collectors.toList());

        assertEquals("ARDOUGNE", events.get(0).diaryArea);
        assertEquals("EASY", events.get(0).tier);
        assertEquals("Enter the Wilderness", events.get(0).taskName);
        assertEquals(false, events.get(0).completed);

        assertEquals("EASY", events.get(1).tier);
        assertEquals("Talk to the Ardougne baker", events.get(1).taskName);
        assertEquals(true, events.get(1).completed);

        assertEquals("MEDIUM", events.get(2).tier);
        assertEquals("Steal from the Ardougne market stalls", events.get(2).taskName);
        assertEquals(false, events.get(2).completed);
    }

    @Test
    public void ignoresWidgetLoadsFromInterfacesOtherThanTheJournalScroll() {
        setPage("Ardougne Area Tasks", "Easy", "Enter the Wilderness");
        tracker.onWidgetLoaded(loadEvent(999));
        verify(apiClient, never()).enqueue(any());
    }

    @Test
    public void ignoresJournalScrollPagesThatAreNotAnAchievementDiaryEgTheQuestJournal() {
        Widget titleWidget = widget("Quest Journal");
        when(client.getWidget(InterfaceID.Journalscroll.TITLE)).thenReturn(titleWidget);

        Widget[] lineWidgets = new Widget[] { widget("Dragon Slayer II"), widget("Some quest step") };
        Widget textLayer = mock(Widget.class);
        when(textLayer.getStaticChildren()).thenReturn(lineWidgets);
        when(client.getWidget(InterfaceID.Journalscroll.TEXTLAYER)).thenReturn(textLayer);

        tracker.onWidgetLoaded(loadEvent());

        verify(apiClient, never()).enqueue(any());
    }

    @Test
    public void resolvesCompoundAreaHeadingsToTheSameAreaTokenTheTierCompleteTrackerUses() {
        setPage("Kourend & Kebos Tasks", "Elite", "Do a hard Kourend thing");
        tracker.onWidgetLoaded(loadEvent());

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        assertEquals("KOUREND", ((StatEvent.DiaryTaskProgress) captor.getValue()).diaryArea);
    }

    @Test
    public void anUnrecognizedAreaHeadingEmitsNothingRatherThanGuessing() {
        setPage("Some Unknown Page Heading", "Easy", "A task");
        tracker.onWidgetLoaded(loadEvent());
        verify(apiClient, never()).enqueue(any());
    }

    @Test
    public void taskLinesBeforeTheFirstTierHeaderAreSkippedNotMisattributed() {
        setPage("Varrock Tasks", "A stray line with no tier context yet", "Easy", "A real task");

        tracker.onWidgetLoaded(loadEvent());

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        assertEquals("A real task", ((StatEvent.DiaryTaskProgress) captor.getValue()).taskName);
    }

    @Test
    public void blankLinesAreSkippedWithoutBreakingTierTracking() {
        setPage("Varrock Tasks", "Easy", "", "First task", "", "Second task");

        tracker.onWidgetLoaded(loadEvent());

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient, times(2)).enqueue(captor.capture());
        List<StatEvent.DiaryTaskProgress> events = captor.getAllValues().stream()
            .map(e -> (StatEvent.DiaryTaskProgress) e)
            .collect(Collectors.toList());
        assertEquals("EASY", events.get(0).tier);
        assertEquals("EASY", events.get(1).tier);
    }
}

File: plugin/src/test/kotlin/com/oldstats/tracking/DiaryTaskTrackerTest.kt
package com.oldstats.tracking

import com.oldstats.api.OldStatsApiClient
import com.oldstats.api.StatEvent
import net.runelite.api.Client
import net.runelite.api.events.WidgetLoaded
import net.runelite.api.gameval.InterfaceID
import net.runelite.api.widgets.Widget
import net.runelite.client.callback.ClientThread
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.whenever

/**
 * [DiaryTaskTracker] reads the diary journal's "journal scroll" interface
 * (group [InterfaceID.JOURNALSCROLL]) rather than a varbit, so these tests
 * build fake title/text-layer widgets the same shape RuneLite's own
 * decompiled `DiaryRequirementsPlugin` reads (confirmed via `cfr`).
 */
class DiaryTaskTrackerTest {
    private lateinit var apiClient: OldStatsApiClient
    private lateinit var client: Client
    private lateinit var clientThread: ClientThread
    private lateinit var tracker: DiaryTaskTracker

    @Before
    fun setUp() {
        apiClient = mock()
        client = mock()
        clientThread = mock()
        tracker = DiaryTaskTracker(apiClient, client, clientThread)

        // Run the deferred client-thread callback synchronously so tests can assert immediately.
        doAnswer { invocation -> (invocation.getArgument(0) as Runnable).run() }
            .whenever(clientThread)
            .invokeLater(any<Runnable>())
    }

    private fun widget(text: String): Widget {
        val w: Widget = mock()
        whenever(w.text).thenReturn(text)
        return w
    }

    private fun setPage(heading: String, vararg lines: String) {
        val titleWidget = widget("Achievement Diary")
        whenever(client.getWidget(InterfaceID.Journalscroll.TITLE)).thenReturn(titleWidget)

        val lineWidgets = (arrayOf(heading) + lines).map { widget(it) }.toTypedArray()
        val textLayer: Widget = mock()
        whenever(textLayer.staticChildren).thenReturn(lineWidgets)
        whenever(client.getWidget(InterfaceID.Journalscroll.TEXTLAYER)).thenReturn(textLayer)
    }

    private fun loadEvent(groupId: Int = InterfaceID.JOURNALSCROLL): WidgetLoaded {
        val event = WidgetLoaded()
        event.groupId = groupId
        return event
    }

    @Test
    fun `emits one event per task line with the correct tier and completion state`() {
        setPage(
            "Ardougne Area Tasks",
            "Easy",
            "Enter the Wilderness",
            "<str>Talk to the Ardougne baker</str>",
            "Medium",
            "Steal from the Ardougne market stalls",
        )

        tracker.onWidgetLoaded(loadEvent())

        val captor = argumentCaptor<StatEvent>()
        org.mockito.kotlin.verify(apiClient, org.mockito.kotlin.times(3)).enqueue(captor.capture())
        val events = captor.allValues.map { it as StatEvent.DiaryTaskProgress }

        assertEquals("ARDOUGNE", events[0].diaryArea)
        assertEquals("EASY", events[0].tier)
        assertEquals("Enter the Wilderness", events[0].taskName)
        assertEquals(false, events[0].completed)

        assertEquals("EASY", events[1].tier)
        assertEquals("Talk to the Ardougne baker", events[1].taskName)
        assertEquals(true, events[1].completed)

        assertEquals("MEDIUM", events[2].tier)
        assertEquals("Steal from the Ardougne market stalls", events[2].taskName)
        assertEquals(false, events[2].completed)
    }

    @Test
    fun `ignores widget loads from interfaces other than the journal scroll`() {
        setPage("Ardougne Area Tasks", "Easy", "Enter the Wilderness")
        tracker.onWidgetLoaded(loadEvent(groupId = 999))
        org.mockito.kotlin.verify(apiClient, never()).enqueue(any())
    }

    @Test
    fun `ignores journal scroll pages that are not an achievement diary (e g the quest journal)`() {
        val titleWidget = widget("Quest Journal")
        whenever(client.getWidget(InterfaceID.Journalscroll.TITLE)).thenReturn(titleWidget)

        val lineWidgets = arrayOf(widget("Dragon Slayer II"), widget("Some quest step"))
        val textLayer: Widget = mock()
        whenever(textLayer.staticChildren).thenReturn(lineWidgets)
        whenever(client.getWidget(InterfaceID.Journalscroll.TEXTLAYER)).thenReturn(textLayer)

        tracker.onWidgetLoaded(loadEvent())

        org.mockito.kotlin.verify(apiClient, never()).enqueue(any())
    }

    @Test
    fun `resolves compound area headings to the same area token the tier-complete tracker uses`() {
        setPage("Kourend & Kebos Tasks", "Elite", "Do a hard Kourend thing")
        tracker.onWidgetLoaded(loadEvent())

        val captor = argumentCaptor<StatEvent>()
        org.mockito.kotlin.verify(apiClient).enqueue(captor.capture())
        assertEquals("KOUREND", (captor.firstValue as StatEvent.DiaryTaskProgress).diaryArea)
    }

    @Test
    fun `an unrecognized area heading emits nothing rather than guessing`() {
        setPage("Some Unknown Page Heading", "Easy", "A task")
        tracker.onWidgetLoaded(loadEvent())
        org.mockito.kotlin.verify(apiClient, never()).enqueue(any())
    }

    @Test
    fun `task lines before the first tier header are skipped, not misattributed`() {
        setPage("Varrock Tasks", "A stray line with no tier context yet", "Easy", "A real task")

        tracker.onWidgetLoaded(loadEvent())

        val captor = argumentCaptor<StatEvent>()
        org.mockito.kotlin.verify(apiClient).enqueue(captor.capture())
        assertEquals("A real task", (captor.firstValue as StatEvent.DiaryTaskProgress).taskName)
    }

    @Test
    fun `blank lines are skipped without breaking tier tracking`() {
        setPage("Varrock Tasks", "Easy", "", "First task", "", "Second task")

        tracker.onWidgetLoaded(loadEvent())

        val captor = argumentCaptor<StatEvent>()
        org.mockito.kotlin.verify(apiClient, org.mockito.kotlin.times(2)).enqueue(captor.capture())
        val events = captor.allValues.map { it as StatEvent.DiaryTaskProgress }
        assertEquals("EASY", events[0].tier)
        assertEquals("EASY", events[1].tier)
    }
}
